### ISSUE TITLE
Evaluate stability of Zeebe with high latency between regions

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#16768]
Goals:
* Finding a max RTT where Zeebe can run without any stability issues
* Find relevant configuration changes
* How exporting to ElasticSearch is affected by high latency

Out of scope:
* It is expected that with high latency, performance will be impacted significantly. Performance improvements are out of scope.

**Comment by deepthidevaki** [comment]
To evaluate the impact of latency, we setup a 2-region cluster using our benchmark helm. Each region is a different namespace. It is setup using `region-0.sh` and `region-1.sh` given [here](https://github.com/camunda/zeebe/tree/dd-evaluate-2-region-latency/2-region-benchmark). 

Following parameters are configured to address high latency
```
 zeebe:
    env:
    - name: ZEEBE_BROKER_CLUSTER_INITIALCONTACTPOINTS
      value: "camunda-zeebe-0.camunda-zeebe.dd-region-0.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.dd-region-0.svc.cluster.local:26502,camunda-zeebe-0.camunda-zeebe.dd-region-1.svc.cluster.local:26502,camunda-zeebe-1.camunda-zeebe.dd-region-1.svc.cluster.local:26502"
    - name: ZEEBE_BROKER_CLUSTER_MESSAGECOMPRESSION
      value: "GZIP"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_REQUESTTIMEOUT
      value: "1s"
    - name: ZEEBE_BROKER_CLUSTER_MEMBERSHIP_PROBETIMEOUT
      value: "500ms"
    - name: ZEEBE_BROKER_CLUSTER_MEMBERSHIP_PROBEINTERVAL
      value: "2s"
    - name: ZEEBE_BROKER_EXPERIMENTAL_RAFT_SNAPSHOTREQUESTTIMEOUT
      value: "10s"

  zeebe-gateway:
    env:
    - name: ZEEBE_GATEWAY_CLUSTER_MESSAGECOMPRESSION
      value: "GZIP"
    - name: ZEEBE_GATEWAY_CLUSTER_MEMBERSHIP_PROBETIMEOUT
      value: "500ms"
    - name: ZEEBE_GATEWAY_CLUSTER_MEMBERSHIP_PROBEINTERVAL
      value: "2s"
```

Latency between 2 regions is injected using chaos-mesh. The experiment is defined [here](https://github.com/camunda/zeebe/blob/dd-evaluate-2-region-latency/2-region-benchmark/chaos-network-region.yaml)

------
#### Results with 100ms RTT between regions
**TLDR:**

- Everything is slower
- But the cluster is stable with no leader changes
- Snapshot replication takes upto 2m when snapshot size is around 1GB. But it can complete with the above updated configuration. 


Generally everything is slower. This is expected as the commit latency is higher.
![image](https://github.com/camunda/zeebe/assets/1997478/55c0bf76-c69a-49e8-a73c-374e125193f1)
![image](https://github.com/camunda/zeebe/assets/1997478/f0f2a6cc-184a-4ff0-b3d8-c0aebd021bd6)

:heavy_check_mark:  But the cluster is running stable with no leader changes
![image](https://github.com/camunda/zeebe/assets/1997478/9f6ed4f1-1215-44aa-9fe6-b41ae186f828)

This cluster has large snapshot of around 1GB per partition. 
:information_source: Snapshot replication takes almost 2min. 
![image](https://github.com/camunda/zeebe/assets/1997478/d67b18da-a388-4b8f-9869-e6af9d0cad81)

:heavy_check_mark: But the good thing is that snapshot replication can complete with the modified configurations. With the default values, it was not able to complete it and got caught in a retry loop.

**Comment by deepthidevaki** [comment]
#### RTT 100ms
After setting the following parameter (default value is 2)
```
    - name: ZEEBE_BROKER_EXPERIMENTAL_MAXAPPENDSPERFOLLOWER
      value: "12"
```
We can see the throughput is much better eventhough commit latency is still high. 
![image](https://github.com/camunda/zeebe/assets/1997478/3feaf0dc-5285-41fc-8588-aafe473ef46e)
![image](https://github.com/camunda/zeebe/assets/1997478/f1bfd6ad-2afc-4ea0-9bdc-667926593a43)
![image](https://github.com/camunda/zeebe/assets/1997478/a41b53fe-4202-41a3-b566-08e31ce03014)

With default value of 2, throughput was around 20 PI/s

Setting ZEEBE_BROKER_EXPERIMENTAL_MAXAPPENDSPERFOLLOWER to 32, we are able to achieve 150 PI/s. 

![image](https://github.com/camunda/zeebe/assets/1997478/76b3155b-b56f-46b6-85b5-006a2c2b69c2)
![image](https://github.com/camunda/zeebe/assets/1997478/903292d1-fb36-4784-b6e1-e2f6a1ff71d7)

![image](https://github.com/camunda/zeebe/assets/1997478/b8fd31b2-0494-479b-b62c-461c38538b90)

**Comment by deepthidevaki** [comment]
### RTT 200ms

With `ZEEBE_BROKER_EXPERIMENTAL_MAXAPPENDSPERFOLLOWER` to 32

![image](https://github.com/camunda/zeebe/assets/1997478/926ebd83-88a2-4978-a930-1a904602e8d6)

![image](https://github.com/camunda/zeebe/assets/1997478/7d78c4b8-fb6c-4aba-881c-2e02277d4be4)

**Comment by deepthidevaki** [comment]
**TLDR:**
Exporting was fast initially, but started lagging behind. Not clear if it is due to high latency or limited CPU availability.

### Impact of high latency on Exporters

For 2-region, we will have two ElasticSearch - one in each region. Zeebe brokers in both regions exports to two ElasticSearch. The local one has low latency. Exporting to the remote one can be impacted by higher latency.

#### With no latency between regions

![image](https://github.com/camunda/zeebe/assets/1997478/0ab628f4-bdd3-49e5-a032-7ee4ef352b03)

#### With 100ms latency between regions

![image](https://github.com/camunda/zeebe/assets/1997478/70f3b01b-69f3-4ad4-974a-fe5b2041aa44)

Initially, except for 1 partition exporting seems to be fast enough. Later there was a pod restart which resulted in imbalanced leadership. From that moment onwards the exporting in 2 partitions started lagging behind. This is also reflected in the throughput acheived. Later the load was reduced from 150 PI/s to 100 PI and the exporting started to catch up slowly. 

It is not clear if the lagging was a result of higher network latency or limited CPU availability. With two leaders on same broker, and exporting to two ES per partition it might need more CPU to reach the target exporting throughput.

**Comment by deepthidevaki** [comment]
Summary:

- Zeebe is running stable with high RTT of 100ms and 200ms.
- Exporting is also running fine. It is sometimes lagging. However, it seems to be because of CPU contention rather than higher latency. 
- With some tuning of configuration, throughput is also fine. Multi-region with 100ms RTT is able to achieve a throughput of 150 PI/s which is same as our default benchmark setup with no latency :tada:. Overall process instance duration is higher, as expected.
- The main issue is snapshot replication which can take upto several minutes depending on the size. With 100ms RTT it is taking upto 2 minutes. With 200ms RTT, it is upto 3 minutes for a snapshot of size ~ 1GB.

**Comment by deepthidevaki** [comment]
Based on above results, I think we can recommend in general, a max RTT of 100ms. Snapshot replication is already impacted at this latency. Anything higher would have a more noticeable impact on snapshot replication and thus availability. Zeebe would also work with higher latency, but it should be evaluated case by case depending on the workload.  What do you think @npepinpe 

Following configuration changes are recommended:

```
 zeebe:
    env:
    - name: ZEEBE_BROKER_CLUSTER_MESSAGECOMPRESSION
      value: "GZIP"
    - name: ZEEBE_BROKER_BACKPRESSURE_AIMD_REQUESTTIMEOUT
      value: "1s"
    - name: ZEEBE_BROKER_CLUSTER_MEMBERSHIP_PROBETIMEOUT
      value: "500ms"
    - name: ZEEBE_BROKER_CLUSTER_MEMBERSHIP_PROBEINTERVAL
      value: "2s"
    - name: ZEEBE_BROKER_EXPERIMENTAL_RAFT_SNAPSHOTREQUESTTIMEOUT
      value: "10s"
    - name: ZEEBE_BROKER_EXPERIMENTAL_MAXAPPENDSPERFOLLOWER
      value: "32"
    - name: ZEEBE_BROKER_EXPERIMENTAL_RAFT_PREFERSNAPSHOTREPLICATIONTHRESHOLD
      value: "10000"

  zeebe-gateway:
    env:
    - name: ZEEBE_GATEWAY_CLUSTER_MESSAGECOMPRESSION
      value: "GZIP"
    - name: ZEEBE_GATEWAY_CLUSTER_MEMBERSHIP_PROBETIMEOUT
      value: "500ms"
    - name: ZEEBE_GATEWAY_CLUSTER_MEMBERSHIP_PROBEINTERVAL
      value: "2s"
```

**Comment by npepinpe** [comment]
Sounds good. As we have a REST API now, I would also recommend adding the following property for versions >= 8.5:

```properties
server.compression.enabled=true
```
